src/approval/ApprovalController.ts

```diff
@@ -34,29 +39,107 @@ export interface Approval {
   /**
    * The type of the approval request.
    */
-  type: string;
+  type: Type;
 
   /**
    * Additional data associated with the request.
+   * TODO:TS4.4 make optional
    */
-  requestData?: RequestData;
-}
+  requestData: RequestData;
+};
 
-export interface ApprovalConfig extends BaseConfig {
-  showApprovalRequest: () => void;
-}
+type ShowApprovalRequest = () => void | Promise<void>;
 
-export interface ApprovalState extends BaseState {
-  [APPROVALS_STORE_KEY]: { [approvalId: string]: Approval };
-  [APPROVAL_COUNT_STORE_KEY]: number;
-}
+export type ApprovalControllerState = {
+  pendingApprovals: Record<
+    string,
+    ApprovalRequest<string, Record<string, Json>>
+  >;
+  pendingApprovalCount: number;
+};
+
+const stateMetadata = {
+  pendingApprovals: { persist: false, anonymous: true },
+  pendingApprovalCount: { persist: false, anonymous: false },
+};
 
 const getAlreadyPendingMessage = (origin: string, type: string) =>
   `Request of type '${type}' already pending for origin ${origin}. Please wait.`;
 
-const defaultState: ApprovalState = {
-  [APPROVALS_STORE_KEY]: {},
-  [APPROVAL_COUNT_STORE_KEY]: 0,
+const getDefaultState = (): ApprovalControllerState => {
+  return {
+    pendingApprovals: {},
+    pendingApprovalCount: 0,
+  };
+};
+
+export type GetApprovalsState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => ApprovalControllerState;
+};
+
+export type ClearApprovalRequests = {
+  type: `${typeof controllerName}:clearRequests`;
+  handler: () => void;
+};
+
+type AddApprovalOptions<RequestData extends Record<string, Json>> = {
+  id?: string;
+  origin: string;
+  type: string;
+  requestData?: RequestData;
```

Comment:
Why use a generic here, rather than `Record<string, Json>`? Is this ever set to anything other than that?

I can see it's passed through from the functions `add` and `addAndShowApprovalRequest`, but it doesn't seem like it's adding any additional type safety, because it's never set to the specific shape expected for a given approval request.

If you want to ensure type safety for the request data, I think you'll need to make the entire class take a generic which is the set of all approval requests, and then check that the type matches the approval options shape using that.

Reply:
Yeah, after considering this, I decided to remove all useless generics rather than going through the trouble of creating that all-ecompassing approval union generic. For now.